File: src-tauri/src/updater.rs
use log::{info, warn, error};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, Manager};
use tauri_plugin_updater::UpdaterExt;

/// Information about an available update
#[derive(Clone, Serialize, Deserialize)]
pub struct UpdateInfo {
    pub version: String,
    pub current_version: String,
    pub body: Option<String>,
    pub date: Option<String>,
}

/// Progress information during download
#[derive(Clone, Serialize, Deserialize)]
pub struct UpdateProgress {
    pub downloaded: u64,
    pub total: Option<u64>,
    pub percent: Option<f64>,
}

/// Check for updates and return info if available
#[tauri::command]
pub async fn check_for_updates(app: AppHandle) -> Result<Option<UpdateInfo>, String> {
    info!("Checking for updates...");
    
    let updater = app.updater().map_err(|e| {
        error!("Failed to get updater: {}", e);
        e.to_string()
    })?;
    
    match updater.check().await {
        Ok(Some(update)) => {
            let current_version = env!("CARGO_PKG_VERSION").to_string();
            let info = UpdateInfo {
                version: update.version.clone(),
                current_version,
                body: update.body.clone(),
                date: update.date.map(|d| d.to_string()),
            };
            info!("Update available: {} -> {}", info.current_version, info.version);
            Ok(Some(info))
        }
        Ok(None) => {
            info!("No updates available");
            Ok(None)
        }
        Err(e) => {
            warn!("Failed to check for updates: {}", e);
            Err(e.to_string())
        }
    }
}

/// Download and install the available update
#[tauri::command]
pub async fn download_and_install_update(app: AppHandle) -> Result<(), String> {
    info!("Starting update download...");
    
    let updater = app.updater().map_err(|e| e.to_string())?;
    
    let update = updater.check().await
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "No update available".to_string())?;
    
    let app_handle = app.clone();
    
    // Download with progress reporting
    let mut downloaded = 0u64;
    let total = update.download_size;
    
    let bytes = update.download(
        |chunk_len, content_len| {
            downloaded += chunk_len as u64;
            let progress = UpdateProgress {
                downloaded,
                total: content_len,
                percent: content_len.map(|t| (downloaded as f64 / t as f64) * 100.0),
            };
            // Emit progress event to frontend
            let _ = app_handle.emit("update:progress", progress);
        },
        || {
            // Download complete callback
            info!("Update download complete");
        },
    ).await.map_err(|e| {
        error!("Failed to download update: {}", e);
        e.to_string()
    })?;
    
    info!("Downloaded {} bytes, installing...", bytes.len());
    
    // Emit installing event
    let _ = app.emit("update:installing", ());
    
    // Install the update - this will restart the app
    update.install(bytes).map_err(|e| {
        error!("Failed to install update: {}", e);
        e.to_string()
    })?;
    
    // Request app restart
    info!("Update installed, restarting...");
    app.restart();
}

/// Check for updates on startup (silent check, only notifies if update available)
pub async fn check_updates_on_startup(app: AppHandle) {
    // Wait a few seconds after startup before checking
    tokio::time::sleep(std::time::Duration::from_secs(3)).await;
    
    info!("Performing startup update check...");
    
    let updater = match app.updater() {
        Ok(u) => u,
        Err(e) => {
            warn!("Updater not available: {}", e);
            return;
        }
    };
    
    match updater.check().await {
        Ok(Some(update)) => {
            let current_version = env!("CARGO_PKG_VERSION").to_string();
            let info = UpdateInfo {
                version: update.version.clone(),
                current_version,
                body: update.body.clone(),
                date: update.date.map(|d| d.to_string()),
            };
            
            info!("Update available on startup: {}", info.version);
            
            // Emit event to frontend
            let _ = app.emit("update:available", info);
        }
        Ok(None) => {
            info!("No updates available on startup check");
        }
        Err(e) => {
            // Silent failure on startup - don't bother user
            warn!("Startup update check failed: {}", e);
        }
    }
}

File: src/lib/stores/updater.ts
import { writable, derived } from 'svelte/store';
import { browser } from '$app/environment';

export type UpdateStatus =
	| 'idle'
	| 'checking'
	| 'available'
	| 'downloading'
	| 'installing'
	| 'up-to-date'
	| 'error';

export interface UpdateInfo {
	version: string;
	current_version: string;
	body: string | null;
	date: string | null;
}

export interface UpdateProgress {
	downloaded: number;
	total: number | null;
	percent: number | null;
}

export interface UpdaterState {
	status: UpdateStatus;
	updateInfo: UpdateInfo | null;
	progress: UpdateProgress | null;
	error: string | null;
	lastChecked: number | null;
	dismissed: boolean;
}

const initialState: UpdaterState = {
	status: 'idle',
	updateInfo: null,
	progress: null,
	error: null,
	lastChecked: null,
	dismissed: false
};

function createUpdaterStore() {
	const { subscribe, set, update } = writable<UpdaterState>(initialState);
	let listenersSetup = false;

	async function setupListeners() {
		if (!browser || listenersSetup) return;
		listenersSetup = true;

		try {
			const { listen } = await import('@tauri-apps/api/event');

			// Listen for startup update check results
			await listen<UpdateInfo>('update:available', (event) => {
				update((s) => ({
					...s,
					status: 'available',
					updateInfo: event.payload,
					dismissed: false,
					error: null
				}));
			});

			// Listen for download progress
			await listen<UpdateProgress>('update:progress', (event) => {
				update((s) => ({
					...s,
					status: 'downloading',
					progress: event.payload
				}));
			});

			// Listen for install phase
			await listen('update:installing', () => {
				update((s) => ({
					...s,
					status: 'installing',
					progress: null
				}));
			});
		} catch {
			// Not in Tauri environment
		}
	}

	async function checkForUpdates(): Promise<UpdateInfo | null> {
		update((s) => ({ ...s, status: 'checking', error: null }));

		try {
			const { invoke } = await import('@tauri-apps/api/core');
			const result = await invoke<UpdateInfo | null>('check_for_updates');

			if (result) {
				update((s) => ({
					...s,
					status: 'available',
					updateInfo: result,
					lastChecked: Date.now(),
					dismissed: false,
					error: null
				}));
				return result;
			} else {
				update((s) => ({
					...s,
					status: 'up-to-date',
					updateInfo: null,
					lastChecked: Date.now(),
					error: null
				}));
				return null;
			}
		} catch (err) {
			const message = err instanceof Error ? err.message : String(err);
			update((s) => ({
				...s,
				status: 'error',
				error: message,
				lastChecked: Date.now()
			}));
			return null;
		}
	}

	async function downloadAndInstall(): Promise<void> {
		update((s) => ({ ...s, status: 'downloading', progress: null, error: null }));

		try {
			const { invoke } = await import('@tauri-apps/api/core');
			await invoke('download_and_install_update');
			// App will restart after install, so we won't reach here normally
		} catch (err) {
			const message = err instanceof Error ? err.message : String(err);
			update((s) => ({
				...s,
				status: 'error',
				error: message
			}));
		}
	}

	function dismiss() {
		update((s) => ({ ...s, dismissed: true }));
	}

	function reset() {
		set(initialState);
	}

	// Setup listeners on creation
	if (browser) {
		setupListeners();
	}

	return {
		subscribe,
		checkForUpdates,
		downloadAndInstall,
		dismiss,
		reset
	};
}

export const updater = createUpdaterStore();

export const updateStatus = derived(updater, ($u) => $u.status);
export const updateAvailable = derived(
	updater,
	($u) => $u.status === 'available' && !$u.dismissed
);
export const updateProgress = derived(updater, ($u) => $u.progress);
export const updateInfo = derived(updater, ($u) => $u.updateInfo);
